### Note
* Instructions have been included for each segment. You do not have to follow them exactly, but they are included to help you think through the steps.

In [25]:
# Dependencies and Setup
import pandas as pd
import numpy as np

In [26]:
# File to Load
file = "C:/Users/Melissa.memel/Desktop/DATA ANALYTICS BC/Homework/Pandas-Challenge/Resources/purchase_data.csv"

In [27]:
# Read Purchasing File and store into Pandas data frame
data_file = pd.read_csv(file)
data_file_df = pd.DataFrame(data_file)
data_file_df.head()

Purchase ID             SN  Age Gender  Item ID  \
0            0        Lisim78   20   Male      108   
1            1    Lisovynya38   40   Male      143   
2            2     Ithergue48   24   Male       92   
3            3  Chamassasya86   24   Male      100   
4            4      Iskosia90   23   Male      131   

                                   Item Name  Price  
0  Extraction, Quickblade Of Trembling Hands   3.53  
1                          Frenzied Scimitar   1.56  
2                               Final Critic   4.88  
3                                Blindscythe   3.27  
4                                       Fury   1.44

## Player Count

* Display the total number of players


In [28]:
total_players = len(data_file["SN"].unique())
total_players_df = pd.DataFrame({"Total Number of Players": [total_players]})
total_players_df

Total Number of Players
0                      576

## Purchasing Analysis (Total)

* Run basic calculations to obtain number of unique items, average price, etc.


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame


In [29]:
# Basic Calculations
unique_items = len(data_file_df["Item ID"].unique())
total_purchases = data_file_df["Purchase ID"].count()
total_revenue = data_file_df["Price"].sum()
average_price = data_file_df["Price"].mean()

In [30]:
# Summary Of Data Frame and Format
purchase_analysis_df = pd.DataFrame([{"Number of Unique Items": unique_items, "Average Price": average_price,
                                      "Number of Purchases": total_purchases, "Total Revenue": total_revenue}])
purchase_analysis_df["Average Price"] = purchase_analysis_df["Average Price"].map("${:,.2f}".format)
purchase_analysis_df["Total Revenue"] = purchase_analysis_df["Total Revenue"].map("${:,.2f}".format)
purchase_analysis_df

Number of Unique Items Average Price  Number of Purchases Total Revenue
0                     179         $3.05                  780     $2,379.77

## Gender Demographics

* Percentage and Count of Male Players


* Percentage and Count of Female Players


* Percentage and Count of Other / Non-Disclosed




In [31]:
# Group data by Gender
gender_stats = data_file.groupby("Gender")

# Count the total of "SN" by gender
total_count_gender = gender_stats.nunique()["SN"]

# Percentage Caluculation 
percentage_of_players = total_count_gender / total_players * 100

# Summary of Data Frame , and Formatting
gender_demographics_df = pd.DataFrame({"Percentage of Players": percentage_of_players, "Total Count": total_count_gender})
gender_demographics_df["Percentage of Players"] = gender_demographics_df["Percentage of Players"].map("{:,.2f}%".format)
gender_demographics_df.index.name = None

gender_demographics_df.sort_values(["Total Count"], ascending = False)


Percentage of Players  Total Count
Male                                 84.03%          484
Female                               14.06%           81
Other / Non-Disclosed                 1.91%           11


## Purchasing Analysis (Gender)

* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. by gender




* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [32]:
# basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. by gender
purchase_count = gender_stats["Purchase ID"].count()
avg_purchase_price = gender_stats["Price"].mean()
avg_purchase_total = gender_stats["Price"].sum()
avg_purchase_per_person = avg_purchase_total/total_count_gender

# Summary data frame 
gender_demographics_df = pd.DataFrame({"Purchase Count": purchase_count, 
                                    "Average Purchase Price": avg_purchase_price,
                                    "Average Purchase Value":avg_purchase_total,
                                    "Avg Purchase Total per Person": avg_purchase_per_person})

gender_demographics_df.index.name = "Gender"

# Formatting
gender_demographics_df["Purchase Count"] = gender_demographics_df["Purchase Count"].map("${}".format)
gender_demographics_df["Average Purchase Price"] = gender_demographics_df["Average Purchase Price"].map("${:,.2f}".format)
gender_demographics_df["Average Purchase Value"] = gender_demographics_df["Average Purchase Value"].map("${:,.2f}".format)
gender_demographics_df["Avg Purchase Total per Person"] = gender_demographics_df["Avg Purchase Total per Person"].map("${:,.2f}".format)

gender_demographics_df

Purchase Count Average Purchase Price  \
Gender                                                        
Female                          $113                  $3.20   
Male                            $652                  $3.02   
Other / Non-Disclosed            $15                  $3.35   

                      Average Purchase Value Avg Purchase Total per Person  
Gender                                                                      
Female                               $361.94                         $4.47  
Male                               $1,967.64                         $4.07  
Other / Non-Disclosed                 $50.19                         $4.56

## Age Demographics

* Establish bins for ages


* Categorize the existing players using the age bins. Hint: use pd.cut()


* Calculate the numbers and percentages by age group


* Create a summary data frame to hold the results


* Optional: round the percentage column to two decimal points


* Display Age Demographics Table


In [33]:
# Create bins
age_bins = [0, 9.90, 14.90, 19.90, 24.90, 29.90, 34.90, 39.90, 99999]
group_names = ["<10", "10-14", "15-19", "20-24", "25-29", "30-34", "35-39", "40+"]

# Add bins to new data frame and groupby 
data_file["Age Group"] = pd.cut(data_file["Age"],age_bins, labels=group_names)
age_grouped = data_file.groupby("Age Group")

# Calculate the numbers and percentages by age group
total_count_age = age_grouped["SN"].nunique()
percentage_by_age = (total_count_age/total_players) * 100

# Summary of Data Frame and Formatting
age_demographics_df = pd.DataFrame({"Percentage of Players": percentage_by_age, "Total Count": total_count_age})
age_demographics_df["Percentage of Players"] = age_demographics_df["Percentage of Players"].map("{:,.2f}%".format)

age_demographics_df.index.name = None

age_demographics_df

Percentage of Players  Total Count
<10                   2.95%           17
10-14                 3.82%           22
15-19                18.58%          107
20-24                44.79%          258
25-29                13.37%           77
30-34                 9.03%           52
35-39                 5.38%           31
40+                   2.08%           12

## Purchasing Analysis (Age)

* Bin the purchase_data data frame by age


* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. in the table below


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [34]:
# Basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc.
purchase_count_age = age_grouped["Purchase ID"].count()
avg_purchase_price_age = age_grouped["Price"].mean()
total_purchase_value = age_grouped["Price"].sum() 
avg_purchase_per_person_age = total_purchase_value/total_count_age

# Summary of Data Frame and Formatting
age_demographics_df = pd.DataFrame({"Purchase Count": purchase_count_age,
                                 "Average Purchase Price": avg_purchase_price_age,
                                 "Total Purchase Value":total_purchase_value,
                                 "Average Purchase Total per Person": avg_purchase_per_person_age})

age_demographics_df["Average Purchase Price"] = age_demographics_df["Average Purchase Price"].map("${:,.2f}".format)
age_demographics_df["Total Purchase Value"] = age_demographics_df["Total Purchase Value"].map("${:,.2f}".format)
age_demographics_df["Average Purchase Total per Person"] = age_demographics_df["Average Purchase Total per Person"].map("${:,.2f}".format)
age_demographics_df.index.name = None

age_demographics_df

Purchase Count Average Purchase Price Total Purchase Value  \
<10                23                  $3.35               $77.13   
10-14              28                  $2.96               $82.78   
15-19             136                  $3.04              $412.89   
20-24             365                  $3.05            $1,114.06   
25-29             101                  $2.90              $293.00   
30-34              73                  $2.93              $214.00   
35-39              41                  $3.60              $147.67   
40+                13                  $2.94               $38.24   

      Average Purchase Total per Person  
<10                               $4.54  
10-14                             $3.76  
15-19                             $3.86  
20-24                             $4.32  
25-29                             $3.81  
30-34                             $4.12  
35-39                             $4.76  
40+                               $3.19

## Top Spenders

* Run basic calculations to obtain the results in the table below


* Create a summary data frame to hold the results


* Sort the total purchase value column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [35]:
# Group data by SN
spender_demographics = data_file.groupby("SN")

# Basic calculations to obtain purchase count, avg. purchase price, total purchase value
purchase_count_spender = spender_demographics["Purchase ID"].count()
avg_purchase_price_spender = spender_demographics["Price"].mean() 
total_purchase_spender = spender_demographics["Price"].sum()

# Create data frame, and formatting
top5_spenders_df = pd.DataFrame({"Purchase Count": purchase_count_spender,
                             "Average Purchase Price": avg_purchase_price_spender,
                             "Total Purchase Value":total_purchase_spender})

top5_spenders_format = top5_spenders_df.sort_values(["Total Purchase Value"], ascending=False).head() 

top5_spenders_format["Average Purchase Price"] = top5_spenders_format["Average Purchase Price"].map("${:,.2f}".format)
top5_spenders_format["Total Purchase Value"] = top5_spenders_format["Total Purchase Value"].map("${:,.2f}".format)

top5_spenders_format


Purchase Count Average Purchase Price Total Purchase Value
SN                                                                     
Lisosia93                 5                  $3.79               $18.96
Idastidru52               4                  $3.86               $15.45
Chamjask73                3                  $4.61               $13.83
Iral74                    4                  $3.40               $13.62
Iskadarya95               3                  $4.37               $13.10

## Most Popular Items

* Retrieve the Item ID, Item Name, and Item Price columns


* Group by Item ID and Item Name. Perform calculations to obtain purchase count, item price, and total purchase value


* Create a summary data frame to hold the results


* Sort the purchase count column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [41]:
# Create new data frame with items related information 
items = data_file[["Item ID", "Item Name", "Price"]]

# Group the item data by item id and item name 
item_demographics = items.groupby(["Item ID","Item Name"])

# Calculations to obtain purchase count, item price, and total purchase value 
purchase_count_item = item_demographics["Price"].count() 
purchase_value = (item_demographics["Price"].sum()) 
item_price = purchase_value/purchase_count_item

# Sumary Data Frame and Formatting
top5_items_df = pd.DataFrame({"Purchase Count": purchase_count_item, 
                                   "Item Price": item_price,
                                   "Total Purchase Value":purchase_value})

top5_items_format = top5_items_df.sort_values(["Purchase Count"], ascending=False).head()

top5_items_format["Item Price"] = top5_items_format["Item Price"].map("${:,.2f}".format)
top5_items_format["Total Purchase Value"] = top5_items_format["Total Purchase Value"].map("${:,.2f}".format)

top5_items_format



Purchase Count  \
Item ID Item Name                                                      
92      Final Critic                                              13   
178     Oathbreaker, Last Hope of the Breaking Storm              12   
145     Fiery Glass Crusader                                       9   
132     Persuasion                                                 9   
108     Extraction, Quickblade Of Trembling Hands                  9   

                                                     Item Price  \
Item ID Item Name                                                 
92      Final Critic                                      $4.61   
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
145     Fiery Glass Crusader                              $4.58   
132     Persuasion                                        $3.22   
108     Extraction, Quickblade Of Trembling Hands         $3.53   

                                                     Total Purchase Value  
Item ID Item Name                                                          
92      Final Critic                                               $59.99  
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
145     Fiery Glass Crusader                                       $41.22  
132     Persuasion                                                 $28.99  
108     Extraction, Quickblade Of Trembling Hands                  $31.77

* Sort the above table by total purchase value in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the data frame



In [44]:
# Sort the above table by total purchase value in descending order
top5_items_format = top5_items_df.sort_values(["Total Purchase Value"],
                                                   ascending=False).head()

# Format
top5_items_format.style.format({"Item Price":"${:,.2f}",
                                "Total Purchase Value":"${:,.2f}"})


* Out of the 576 total active players, the main purchasers are male (84%). Female and Other/Non-Disclosed accounts only to (16%) of Total Purchase amount.
* The data shows that approximately 45% of the total customers are in the age range of 20 to 24  years old.
* The item (Final Critic) is the most profitable as well as the most popular items are “Final Critic” and “Oathbreaker, Last Hope of the Breaking Storm”.